src/plugins/discord/address.js

```diff
@@ -0,0 +1,56 @@
+// @flow
+
+import type {NodeAddressT, EdgeAddressT} from "../../core/graph";
+import {NodeAddress, EdgeAddress} from "../../core/graph";
+import {type Snowflake} from "./models";
+import {
+  userNodeType,
+  messageNodeType,
+  sentMessageEdgeType,
+} from "./declaration";
+
+/**
+ * These user IDs are considered globally unique by Discord.
+ * We can reference a user even across multiple guilds using this address.
+ * See https://discordapp.com/api/users/{userId}
+ *
+ * Importantly we don't include any membership details in the address.
+ * This way the graph topology stays identical when users leave / rejoin.
+ */
+export function userAddress(userId: Snowflake): NodeAddressT {
+  return NodeAddress.append(userNodeType.prefix, userId);
+}
```

Comment:
Assuming that a Discord alias looks like `discord/decentralion#1337`, it will be impossible to do address <-> alias mapping without an extra data store, which is sad.

Is it feasible to address users by the name and identifier? I know that these are mutable, which can be a pain, but the same is true for GitHub usernames (i have renamed my account several times). Since we treat the external data store as source of truth, any issues around user id renaming could be resolved by resetting the cache.

Reply:
In the prototype, for that reason the alias uses the user.id Snowflake.
`discord/143776454050709505`

For reference, Discord clients and bots are also expected to use Snowflakes for mentions and channel references if they want them to be connected in a message.
So clients do not send `Hi @Beanow, try posting that in #support`
but `Hi <@143776454050709505>, try posting that in <#454007860926611478>`.

This means that all translation between these is done with local maps and only the IDs are considered valid input on the server. We will have to rely on a cache or API to do this.

---

> Since we treat the external data store as source of truth, any issues around user id renaming could be resolved by resetting the cache.

I don't think this is true when combined with a Grain ledger. In terms of users an entire section of the graph might have a new node that receives it's Cred. So your previous identity will be overpaid and a brand new grain account will be underpaid and receiving that cred.

Having the same situation with GitHub might be more of a reason to change GitHubs node addresses than to do the same thing for Discord. :sweat_smile:

It's worth noting that GitHub also gives very explicit warnings that things are _likely to break_ when you rename things and requires you to confirm that you understand this. Plus most of the time you're talking about a more technical audience than discord users.
![image](https://user-images.githubusercontent.com/497556/82211439-1f7cfc80-9911-11ea-8fce-efb2d90451c8.png)

Worse yet, it seems that Discord offers the ability to control the discriminator (rather than have it randomized) as a premium feature. I don't know if they keep a blacklist of previously used pairs to avoid identity theft, but if they don't we could open ourselves up to trivial identity and Grain theft methods here.

Reply:
Thanks for the explanation @Beanow -- that makes a lot of sense.